src/Expr.h

```diff
@@ -171,21 +171,65 @@ class Expr : public Obj {
 	void MarkParen()		{ paren = true; }
 	bool IsParen() const		{ return paren; }
 
-	const ListExpr* AsListExpr() const;
-	ListExpr* AsListExpr();
+#undef ACCESSORS
+#define ACCESSORS(ctype) \
+	const ctype* As ## ctype () const; \
+	ctype* As ## ctype (); \
+	IntrusivePtr<ctype> As ## ctype ## Ptr ();
+
+	ACCESSORS(ListExpr)
+	ACCESSORS(NameExpr)
+	ACCESSORS(AssignExpr)
+	ACCESSORS(IndexExpr)
+	ACCESSORS(EventExpr)
 
-	const NameExpr* AsNameExpr() const;
-	NameExpr* AsNameExpr();
+	void Describe(ODesc* d) const override final;
 
-	const AssignExpr* AsAssignExpr() const;
-	AssignExpr* AsAssignExpr();
+	virtual TraversalCode Traverse(TraversalCallback* cb) const = 0;
 
-	const IndexExpr* AsIndexExpr() const;
-	IndexExpr* AsIndexExpr();
 
-	void Describe(ODesc* d) const override final;
+	//
+	// BEGINNING of methods relating to script optimization.
+	//
 
-	virtual TraversalCode Traverse(TraversalCallback* cb) const = 0;
+	// Returns a duplicate of the expression.  For atomic expressions
+	// that can be safely shared across multiple function bodies
+	// (due to inline-ing), and that won't have Reaching Definitions
+	// tied to an individual copy, we can return just a reference, per the
+	// default here.
+	virtual ExprPtr Duplicate()	{ return {NewRef{}, this}; }
```

Comment:
> > * got the impression that details are closely tied to how script-optimization logic intends to use it
> 
> I don't think that's the case. The only echoes of this are comments that say "we need to duplicate this node even if you'd think we don't, due to how script optimization associates state with nodes". However, a pure `Duplicate()` method would always duplicate anyway, obviating the need for these comments. Happy to take them out; but I do think `Duplicate()` is the right name, because that's indeed what's happening, the AST is being completely duplicated.

This was the code that specifically confused me and that I still don't know how to integrate with the above explanation:

* Doesn't this show some nodes won't duplicate, just return a reference?
* The comment read to me like "we get away with this since it optimization-specific stuff is not impacted".

Maybe an independent confusion: "atomic expression" didn't have a clear meaning/intent.  My guess was it meant `ConstExpr` and looking for "which subclasses don't override" seems to verify that.  If that's right, may be clearer to have this be pure-virtual in base class and let only `ConstExpr` have the unique override/comments of returning a reference.

Reply:
Thanks for flagging.  On reflection there's no need for the special-casing here, so I've gone ahead regularized this so that it's true-and-full duplication of the AST.  I also got rid of the modifications to `ODesc`.  I'll push shortly to a new standalone branch that doesn't bring in any of the compiler-related stuff.